src/components/validation/runner/telemetry.rs

```diff
@@ -64,30 +64,70 @@ impl Telemetry {
+        // from the Vector sink that we're using to collect telemetry.
+        let grpc_task_coordinator = TaskCoordinator::new();
+        spawn_grpc_server(self.listen_addr, self.service, &grpc_task_coordinator);
+        let grpc_task_coordinator = grpc_task_coordinator.started().await;
+        debug!("All gRPC task(s) started.");
 
         let mut rx = self.rx;
         let driver_handle = tokio::spawn(async move {
             telemetry_started.mark_as_done();
 
             let mut telemetry_events = Vec::new();
-            loop {
+            'outer: loop {
                 select! {
                     _ = telemetry_shutdown_handle.wait() => {
-                        match rx.recv().await {
-                            None => break,
-                            Some(telemetry_event) => telemetry_events.push(telemetry_event),
+                        // After we receive the shutdown signal, we need to wait
+                        // for two event emissions from the internal_metrics
+                        // source. This is to ensure that we've received all the
+                        // events from the components that we're testing.
+                        //
+                        // We need exactly two because the internal_metrics
+                        // source does not emit component events until after the
+                        // component_received_events_total metric has been
+                        // emitted. Thus, two events ensure that all component
+                        // events have been emitted.
+
+                        debug!("telemetry: waiting for final internal_metrics events before shutting down");
+
+                        let mut events_seen = 0;
+                        let current_time = chrono::Utc::now();
+
+                        loop {
+                        match &rx.recv().await {
+                            None => break 'outer,
+                            Some(telemetry_event) => {
+                                    telemetry_events.push(telemetry_event.clone());
+
+                                    if let Event::Metric(metric) = telemetry_event {
+                                        if let Some(tags) = metric.tags() {
+                                            if metric.name() == "component_received_events_total" &&
+                                               tags.get("component_name") == Some(INTERNAL_LOGS_KEY) &&
+                                               metric.data().timestamp().unwrap() > &current_time {
+                                                debug!("telemetry: processed one component_received_events_total event");
+
+                                                events_seen += 1;
+                                                if events_seen == 2 {
+                                                    break 'outer;
+                                                }
```

Comment:
Formatting looks off here, single space indent?

Reply:
There was a space off. We're in the select! macro, which forces me to manually format this section.

Reply:
Doh, I always forget things inside macros don't get auto formatted.